contracts/external/cw-token-swap/src/types.rs

```diff
@@ -0,0 +1,420 @@
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{to_binary, Addr, BankMsg, Binary, Coin, CosmosMsg, Deps, Uint128, WasmMsg};
+
+use crate::ContractError;
+
+// We create empty trait and implement it for the types we want to use
+// so we can create a generic function that only accepts a vec of those types
+trait CompletionMsgsToCosmosMsg {
+    fn into_checked_cosmos_msg(
+        self,
+        deps: Deps,
+        denom: &str,
+    ) -> Result<(Uint128, CosmosMsg), ContractError>;
+}
+
+/// This function returns a vector of CosmosMsgs from the given vector of CompletionMsgs
+/// It verifies the msgs is not empty (else return empty vec)
+/// It verifies the total amount of funds matches the funds sent in all messages
+fn completion_to_cosmos_msgs<T: CompletionMsgsToCosmosMsg>(
+    deps: Deps,
+    msgs: Vec<T>,
+    amount: Uint128,
+    string: &str, // extra data we need (denom or contract addr)
+) -> Result<Vec<CosmosMsg>, ContractError> {
+    if msgs.is_empty() {
+        return Ok(vec![]);
+    }
+
+    let mut total_amount = Uint128::zero();
+    let cosmos_msgs = msgs
+        .into_iter()
+        .map(|msg| {
+            let (amount, cosmos_msg) = msg.into_checked_cosmos_msg(deps, string)?;
+            total_amount += amount;
+            Ok(cosmos_msg)
+        })
+        .collect::<Result<Vec<CosmosMsg>, ContractError>>()?;
+
+    // Verify that total amount of funds matches funds sent in all messages
+    if total_amount != amount {
+        return Err(ContractError::WrongFundsCalculation {});
+    }
+    Ok(cosmos_msgs)
+}
+
+#[cw_serde]
+pub enum NativeSendMsg {
```

Comment:
all of these variants have a `Vec<Coin>` parameter that we want to be sure to validate. currently, we match and validate on each branch:

```
match type {
   a => validate_coins(a.coins),
   b => validate_coins(b.coins),
}
```

what if we moved the amount to a top level field in a struct.

```
struct NativeSend {
  coins,
  msg,
}
```

then, we don't have to remember to validate in each match arm which makes it less likely we forget!

```
validate_coins(coins)
match msg {
  a => ...,
  b => ...,
}
```

Reply:
I wish, but the whole point here is that we take all native msgs we allow as is, with the funds parameter, this also tells us how much exactly to send on each msg.
If we do the above, there is zero verification that the coins they sent in `NativeSend `, is the actual coins they set in each msg, so the check will be worthless because we don't verify the actual sent funds in the msgs.